test/functional/feature_bip8.py

```diff
@@ -0,0 +1,124 @@
+            ['-vbparams=testdummy:@-2:@-2'], # Node 3 has TestDummy inactive, but will be restarted with different params
+        ]
+
+    def setup_network(self):
+        self.setup_nodes()
+
+    def test_height(self, height, status, mine_from, restart):
+        if height > self.height:
+            self.log.info("Test status at height {}...".format(height))
+            while self.height < height:
+                blockhash = self.nodes[mine_from].generate(1)[0]
+                block = self.nodes[mine_from].getblock(blockhash, 0)
+                for node in self.nodes:
+                    node.submitblock(block)
+                    assert_equal(node.getbestblockhash(), blockhash)
+                self.height += 1
+        elif height < self.height:
+            assert mine_from is None
+            self.log.info("Roll back to height {}...".format(height))
+            old_block = self.nodes[0].getblockhash(height + 1)
+            for node in self.nodes:
+                node.invalidateblock(old_block)
+        else:
+            assert mine_from is None
+            assert height == 0 and self.height == 0
+            self.log.info("Test status at genesis...")
+
+        self.height = height
+
+        for (i, node), st in zip(enumerate(self.nodes), status):
+            self.log.debug("Node #{}...".format(i))
+            info = node.getblockchaininfo()
+            assert_equal(info["blocks"], height)
+            if st is None:
+                assert "testdummy" not in info["softforks"]
+            else:
+                assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], st)
+
+        if restart:
+            # Restart this node and check that the status is what we expect
+            self.restart_node(restart.node, restart.extra_args)
+            info = self.nodes[restart.node].getblockchaininfo()
+            assert_equal(info["blocks"], height)
+            if restart.status is None:
+                assert "testdummy" not in info["softforks"]
+            else:
+                assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], restart.status)
+
+            # Now restart again to go back to the original settings
+            self.restart_node(restart.node, self.extra_args[restart.node])
+            info = self.nodes[restart.node].getblockchaininfo()
+            assert_equal(info["blocks"], height)
+            if status[restart.node] is None:
+                assert "testdummy" not in info["softforks"]
+            else:
+                assert_equal(info["softforks"]["testdummy"]["bip8"]["status"], status[restart.node])
+
+    def run_test(self):
+        self.log.info("Checking -vbparams")
+        self.stop_node(3)
+        self.nodes[3].assert_start_raises_init_error(extra_args=["-vbparams=testdummy:@-2:@1"], expected_msg="Error: When one of startheight or timeoutheight is -2, both must be -2")
```

Comment:
There's a lot of repetition here. You could factor this out into a function.

Reply:
It's about as repetitive if this were a function IMO.